overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,220 @@
+package backend
+
+import (
+	"archive/zip"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	archiveName  = "archive.tgz"
+	metadataName = "meta.json"
+	metaHashName = "meta.sha3_384"
+
+	userArchivePrefix = "user/"
+	userArchiveSuffix = ".tgz"
+)
+
+// Iter loops over all valid snapshots in the snapshots directory,
+// applying the given function to each. It is that function's
+// responsibility to close the snapshot it's passed. If the function
+// returns error, iteration is stopped (and if the error isn't EOF,
+// it's returned as the error of the iterator).
+func Iter(ctx context.Context, f func(*Reader) error) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	dir, err := os.Open(dirs.SnapshotDir)
+	if osutil.IsDirNotExist(err) {
+		// no dir -> no snapshots
+		return nil
+	}
+
+	var names []string
+	for err == nil {
+		names, err = dir.Readdirnames(100)
+		for _, name := range names {
+			if err = ctx.Err(); err != nil {
+				break
+			}
+
+			var rsh *Reader
+			// wish dirs had Open() that did openat
+			rsh, err = Open(filepath.Join(dirs.SnapshotDir, name))
+			if err != nil {
+				err = nil
```

Comment:
you're in a loop that ends on error, but this particular error should be ignored (because err on Open means the thing you tried to open wasn't a valid snapshot), so ignore the error and carry on